core/src/Dscp/Util/HasLens.hs

```diff
@@ -0,0 +1,117 @@
+-- | Utilities to work with 'HasLens'.
+
+module Dscp.Util.HasLens
+    ( deriveHasLens
+    , deriveHasLensPath
+    , deriveHasLensPathWith
+    , deriveHasLensDirect
+    , deriveHasLensDirectWith
+    ) where
+
+import Control.Lens (DefName (TopName), LensRules, lensField, lensRules, to, traversed)
+import qualified Data.List as L
+import qualified Data.Text as T
+import qualified Language.Haskell.TH as TH
+import Loot.Base.HasLens (HasLens (..))
+
+-- | Where to get lens to produce 'HasLens' instance.
+data DeriveHasLensType
+    = DeriveViaLensOf
+      -- ^ Using `lensOf` to correspond a field accessor
+    | DeriveViaFieldLens LensRules
+      -- ^ Using lens generated using `makeLensesWith` to correspond a field
+      -- accessor
+
+-- | @ deriveHasLensExts deriveWay [lens1 .. lensN] ctx ctxPart @ traverses all
+-- fields of @ ctxPart @ datatype and for each field of type @ F @ produces
+-- @ instance HasLens ctx F where lensOf = lens1 . .. . lensN . fieldLens @,
+-- where @ lens1 . .. . lensN :: Lens' ctx ctxPart @ holds and
+-- @ fieldLens :: Lens' ctxPart F @ is either @ lensOf \@F @ or a lens
+-- generated with @ makeLenses @, depending on @ deriveWay @ parameter.
+deriveHasLensExt
+    :: DeriveHasLensType
+    -> TH.Q TH.Exp
+    -> TH.Name    -- ^ Name of @H@ type
+    -> TH.Name    -- ^ Name of @M@ type
+    -> TH.Q [TH.Dec]
+deriveHasLensExt deriveWay preLensQ ctxName tyName = do
+    ty <- TH.reify tyName
+    decs <- case ty of
+        TH.TyConI (TH.DataD _ _ _ _ [TH.RecC _ fields] _) -> do
+            let fieldNames = fields ^.. traversed . _1 . to show . to toUnqualifiedName
+            forM fields $ \(quaFieldName, _, fieldTy) -> do
+                let fieldName = toUnqualifiedName (show quaFieldName)
+                let mkLensName rules = toLensName rules fieldNames fieldName
+                mkInstance fieldTy mkLensName
+        _ -> fail "Expected name of a datatype with exactly one constructor"
```

Comment:
This error will also be shown on attempt to apply the function to the datatype which has exactly one constructor, but is not a record (as long as you match only `RecC`). I agree that support for `NormalC` would be cumbersome, but you might make better error message.

Reply:
Good point.